triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala

```diff
@@ -86,34 +89,53 @@ class Server(dar: Option[Dar[(PackageId, Package)]], jdbcConfig: Option[JdbcConf
         // will still call addPackage even if we already fed the
         // package via the callback but this is harmless and not
         // expensive.
+        @scala.annotation.tailrec
         def go(r: Result[Unit]): Unit = r match {
           case ResultDone(()) => ()
           case ResultNeedPackage(pkgId, resume) =>
             go(resume(darMap.get(pkgId)))
           case _ => throw new RuntimeException(s"Unexpected engine result $r")
         }
+
         go(compiledPackages.addPackage(pkgId, pkg))
     }
   }
 
   private def getRunningTrigger(uuid: UUID): RunningTrigger = {
-    triggers.get(uuid).get // TODO: Improve as might throw NoSuchElementException.
+    triggers(uuid) // TODO: Improve as might throw NoSuchElementException.
   }
 
-  private def addRunningTrigger(t: RunningTrigger): Unit = {
-    triggers = triggers + (t.triggerInstance -> t)
-    triggersByToken = triggersByToken + (t.jwt -> (triggersByToken.getOrElse(t.jwt, Set()) + t.triggerInstance))
+  private def addRunningTrigger(t: RunningTrigger): Either[String, Unit] = {
+    triggerDao match {
+      case None =>
+        triggers = triggers + (t.triggerInstance -> t)
+        triggersByToken = triggersByToken + (t.jwt -> (triggersByToken.getOrElse(t.jwt, Set()) + t.triggerInstance))
+        Right(())
+      case Some(dao) =>
+        val insert = dao.transact(TriggerDao.addRunningTrigger(t))
+        Try(insert.unsafeRunSync()) match {
+          case Failure(err) => Left(err.toString)
+          case Success(()) => Right(())
+        }
+    }
   }
 
   private def removeRunningTrigger(t: RunningTrigger): Unit = {
     triggers = triggers - t.triggerInstance
-    triggersByToken = triggersByToken + (t.jwt -> (triggersByToken
-      .get(t.jwt)
-      .get - t.triggerInstance))
+    triggersByToken = triggersByToken + (t.jwt -> (triggersByToken(t.jwt) - t.triggerInstance))
```

Comment:
I guess you are leaving this for a separate PR?

Reply:
Yeah.